From e13ba86143a153a5424d8bac53dc660ae7d2f9b6 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 01 May 2024 12:40:37 +0300
Subject: [PATCH] [FSW-4077] fs1: efuse: volt protection pattern

With this patch, made correction to voltage
protection pattern bit positions and moved the
efuse bits definition to embedded specs.
we dont have a way to inject voltage protection
efuse bits in onix, hence volt prot bits check
moved to platform specific and bypassed.

Change-Id: I442ac09cb69e6753a5bc5bcfbe6a9b0d571641d0
---

diff --git a/arch/fs1/arc/asic.c b/arch/fs1/arc/asic.c
index e29c8c7..46d8e23 100644
--- a/arch/fs1/arc/asic.c
+++ b/arch/fs1/arc/asic.c
@@ -10,6 +10,7 @@
 #include <fs1_reg_map.h>
 #include <fs1_blocks_linux_driver.h>
 #include <pll_control_regs.h>
+#include <dev_efuse.h>
 
 #define BOOTROM_UART_BAUDRATE	115200
 
@@ -52,3 +53,11 @@
 
 	return (uint32_t)out_freq;
 }
+
+bool fuse_valid_volt_prot_pattern(uint32_t config_fuse)
+{
+	if ((config_fuse & EFUSE_SYSCFG_VOLT_PROT_BITS) == FUSE_VOLT_PROT_PATTERN)
+		return true;
+
+	return false;
+}
diff --git a/arch/fs1/arc/dev_efuse.c b/arch/fs1/arc/dev_efuse.c
index 52cef35..6350384 100644
--- a/arch/fs1/arc/dev_efuse.c
+++ b/arch/fs1/arc/dev_efuse.c
@@ -10,7 +10,6 @@
 #include <efuse_masks.h>
 #include <fs1_reg_map.h>
 #include <efuse_regs.h>
-#include <dev_efuse.h>
 #include "delay.h"
 
 void fuse_prog_range_config(void)
@@ -38,11 +37,4 @@
 	regmap_writel(fsm_trigger, IO0_H0_PSOC_EFUSE(mmEFUSE_RD));
 }
 
-bool fuse_valid_volt_prot_pattern(uint32_t config_fuse)
-{
-	if ((config_fuse & FUSE_VOLT_PROT_MASK) == FUSE_VOLT_PROT_PATTERN)
-		return true;
-
-	return false;
-}
-
+__weak bool fuse_valid_volt_prot_pattern(__unused uint32_t config_fuse) { return false; }
diff --git a/arch/fs1/arc/dev_efuse.h b/arch/fs1/arc/dev_efuse.h
index 6a281e2..18b8851 100644
--- a/arch/fs1/arc/dev_efuse.h
+++ b/arch/fs1/arc/dev_efuse.h
@@ -14,8 +14,7 @@
 #define FUSE_DELAY_US				5
 #define FUSE_TIMEOUT_US				500
 #define FUSE_RETRIES				(FUSE_TIMEOUT_US / FUSE_DELAY_US)
-#define FUSE_VOLT_PROT_MASK			0x3C00
-#define FUSE_VOLT_PROT_PATTERN			0x1400
+#define FUSE_VOLT_PROT_PATTERN		0x14000
 
 void fuse_prog_range_config(void);
 void fuse_prog_read_address(void);
diff --git a/arch/fs1/arc/onix.c b/arch/fs1/arc/onix.c
index 375964d..f696abd 100644
--- a/arch/fs1/arc/onix.c
+++ b/arch/fs1/arc/onix.c
@@ -19,3 +19,8 @@
 {
 	return ONIX_REF_CLK;
 }
+
+bool fuse_valid_volt_prot_pattern(__unused uint32_t config_fuse)
+{
+	return true;
+}
diff --git a/arch/fs1/arc/pldm.c b/arch/fs1/arc/pldm.c
index d50a78c..5fbef2f 100644
--- a/arch/fs1/arc/pldm.c
+++ b/arch/fs1/arc/pldm.c
@@ -11,6 +11,7 @@
 #include <fs1_blocks_linux_driver.h>
 #include <pll_control_regs.h>
 #include <uart_regs.h>
+#include <dev_efuse.h>
 
 #define PLDM_UART_REF_CLK	250000000ULL
 #define PLDM_UART_BAUDRATE	300
@@ -41,3 +42,11 @@
 {
 	return PLDM_UART_BAUDRATE;
 }
+
+bool fuse_valid_volt_prot_pattern(uint32_t config_fuse)
+{
+	if ((config_fuse & EFUSE_SYSCFG_VOLT_PROT_BITS) == FUSE_VOLT_PROT_PATTERN)
+		return true;
+
+	return false;
+}
